# Before you use this template

This template is just a recommended template for project Report. It only considers the general type of research in our paper pool. Feel free to edit it to better fit your project. You will iteratively update the same notebook submission for your draft and the final submission. Please check the project rubriks to get a sense of what is expected in the template.

---

# FAQ and Attentions
* Copy and move this template to your Google Drive. Name your notebook by your team ID (upper-left corner). Don't eidt this original file.
* This template covers most questions we want to ask about your reproduction experiment. You don't need to exactly follow the template, however, you should address the questions. Please feel free to customize your report accordingly.
* any report must have run-able codes and necessary annotations (in text and code comments).
* The notebook is like a demo and only uses small-size data (a subset of original data or processed data), the entire runtime of the notebook including data reading, data process, model training, printing, figure plotting, etc,
must be within 8 min, otherwise, you may get penalty on the grade.
  * If the raw dataset is too large to be loaded  you can select a subset of data and pre-process the data, then, upload the subset or processed data to Google Drive and load them in this notebook.
  * If the whole training is too long to run, you can only set the number of training epoch to a small number, e.g., 3, just show that the training is runable.
  * For results model validation, you can train the model outside this notebook in advance, then, load pretrained model and use it for validation (display the figures, print the metrics).
* The post-process is important! For post-process of the results,please use plots/figures. The code to summarize results and plot figures may be tedious, however, it won't be waste of time since these figures can be used for presentation. While plotting in code, the figures should have titles or captions if necessary (e.g., title your figure with "Figure 1. xxxx")
* There is not page limit to your notebook report, you can also use separate notebooks for the report, just make sure your grader can access and run/test them.
* If you use outside resources, please refer them (in any formats). Include the links to the resources if necessary.

# Project Github Repo: https://github.com/linjc16/uiuc-cs598-team120

# 1 Introduction
This is an introduction to your report, you should edit this text/mardown section to compose. In this text/markdown, you should introduce:

*   Background of the problem
  1. what type of problem: disease/readmission/mortality prediction,  feature engineeing, data processing, etc
  2. what is the importance/meaning of solving the problem
  3. what is the difficulty of the problem
  4. the state of the art methods and effectiveness.
*   Paper explanation
  1. what did the paper propose
  2. what is the innovations of the method
  3. how well the proposed method work (in its own metrics)
  4. what is the contribution to the reasearch regime (referring the Background above, how important the paper is to the problem).


## 1.1 Background of the problem
Graph-structured data are ubiquitous in various applications of healthcare domain, such as protein-protein interaction (PPI) prediction and drug-drug interaction (DDI) prediction. Graph Attention Networks (GAT) (Velickovic et al., 2018), as one of the most popular Graph Neural Networks (GNNs), has been widely adopted in these applications due to its ability to effectively model relationships between nodes in a graph. In this project, we are trying to explore the role of GAT in healthcare domains. 

1. what type of problem: disease/readmission/mortality prediction,  feature engineeing, data processing, etc \
**Ans:** Drug-drug Interaction (DDI), Protein-protein Interaction (PPI), and Link Prediciton in PubMed Citation Network. Starting from these three tasks, we explore the potential of GAT in healthcare domains.

2. what is the importance/meaning of solving the problem \
**Ans:** These three tasks are representative in healthcare domains: DDI prediciton aims to identify potentially adverse interactions between drugs, helping to enhance patient safety and treatment effectiveness. PPI prediction focuses on understanding how proteins interact within the human body, which is crucial for developing new therapeutic strategies and understanding biological processes. Finally, link prediction in PubMed Citation Network can uncover hidden relationships between various medical studies, enhancing the synthesis of medical knowledge and fostering new research insights. By addressing these problems using the Graph Attention Networks (GAT), we leverage advanced machine learning techniques to provide more accurate predictions and deeper insights

3. what is the difficulty of the problem \
**Ans:** The difficulty of these problems lies in the following areas: complexity of interactions, sparsity of known interactions, and the need for accurate and generalizable predictive models.

4. the state of the art methods and effectiveness \
**Ans:** Since our project focuses on the role of GAT in healthcare domain applications, this question is not applicable.

## 1.2 Paper explanation
1. what did the paper propose \
**Ans**: The paper introduced Graph Attention Networks (GATs), which utilize masked self-attentional layers to operate on graph-structured data. GATs allow each node to attend over its neighborhoods' features dynamically, which improves flexibility and performance over previous methods that relied on graph convolutions.

2. what is the innovations of the method \
**Ans**: The innovations include the use of attention mechanisms that allow for different weights to be assigned to different nodes in a neighborhood, enhancing model capacity and interpretability. The GATs perform efficiently without needing costly matrix operations and are applicable to both inductive and transductive learning tasks. They also employ multi-head attention to stabilize the learning process.

3. how well the proposed method work (in its own metrics) \
**Ans**: GATs achieved or matched state-of-the-art results (till the date it was published) on several transductive and inductive graph benchmarks, including Cora, Citeseer, Pubmed citation networks, and a protein-protein interaction dataset. Notably, the results demonstrated the effectiveness of attention-based models on graph-structured data, particularly showing a significant improvement in the inductive learning task on the protein-protein interaction dataset.

4. what is the contribution to the reasearch regime (referring the Background above, how important the paper is to the problem). \
**Ans**: The paper contributes significantly to the domain of graph neural networks by providing a more flexible and powerful alternative to previous models like GCNs (Graph Convolutional Networks). By applying GATs in the healthcare domains, the performance of most of tasks with graph structures improved a lot. Specifically, in drug-drug interaction prediction task, each drug is modeled as a molecule graph whose features can be extracted by GAT. For the PPI and link prediciton, both tasks contains networks, also a graph structure that can be fed into GATs.



# 2 Scope of Reproducibility:

List hypotheses from the paper you will test and the corresponding experiments you will run.


1.   Hypothesis 1: attention mechanism greatly improves the performace of GAT for most of the time;
- We will run GAT and GCN (GAT w/o attention) on the three tasks.
2.   Hypothesis 2: multi-head benifits the feature extraction process;
- We will run single-head and multi-head GATs.
3.   Hypothesis 3: one of the drawbacks of GAT is that, the attention mechanism is static instead of dynamic (Brody et al., 2022).
- We will run GATv2 and the original GAT.

<!-- You can insert images in this notebook text, [see this link](https://stackoverflow.com/questions/50670920/how-to-insert-an-inline-image-in-google-colaboratory-from-google-drive) and example below:

![sample_image.png](https://drive.google.com/uc?export=view&id=1g2efvsRJDxTxKz-OY3loMhihrEUdBxbc) -->


# 3 Methodology

This methodology is the core of your project. It consists of run-able codes with necessary annotations to show the expeiment you executed for testing the hypotheses.

The methodology at least contains two subsections **data** and **model** in your experiment.

##  3.1 Data
Data includes raw data (MIMIC III tables), descriptive statistics (our homework questions), and data processing (feature engineering).
  * Source of the data: where the data is collected from; if data is synthetic or self-generated, explain how. If possible, please provide a link to the raw datasets.
  * Statistics: include basic descriptive statistics of the dataset like size, cross validation split, label distribution, etc.
  * Data process: how do you munipulate the data, e.g., change the class labels, split the dataset to train/valid/test, refining the dataset.
  * Illustration: printing results, plotting figures for illustration.
  * You can upload your raw dataset to Google Drive and mount this Colab to the same directory. If your raw dataset is too large, you can upload the processed dataset and have a code to load the processed dataset.

1. Source of the data \
(1) DDI: dataset is from Lin et al. 2023. The dataset is a real-world DDI benchmark dataset constructed from DrugBank. The link to the raw dataset is https://github.com/linjc16/R2-DDI. (2) PPI: dataset is from https://data.dgl.ai/dataset/ppi.zip. Real-word dataset. (3) Link Prediction: https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.datasets.CitationFull.html. Real-word dataset.

2. Statistics: \
(1) DDI: 191k positive triples [drugA, drugB, relation], 1708 drugs, grouped into 86 relations. Training/validation/test sets are 3:1:1. (2) PPI: fully follow the https://data.dgl.ai/dataset/ppi.zip. The detailed statistics can be seen in the following output. (3) Link Prediction: Training/validation/test sets are 0.85:0.05:0.1. The detailed statistics can be seen in the following output.

3. Data process: \
For all the three dataset, all the data are processed and we directly use the data for this project.

4. Illustration: \
See the results for each task below.

## Data 1: Drug-drug Interaction (DDI) Prediciton

In [1]:
import pandas as pd
import os

# dir and function to load raw data
raw_data_dir = 'data/drug_interaction'

def load_data(raw_data_dir):
    # implement this function to load raw data to dataframe/numpy array/tensor
    
    df_ddi_train = pd.read_csv(os.path.join(raw_data_dir, 'ddi_training.csv'))
    df_ddi_val = pd.read_csv(os.path.join(raw_data_dir, 'ddi_validation.csv'))
    df_ddi_test = pd.read_csv(os.path.join(raw_data_dir, 'ddi_test.csv'))

    return df_ddi_train, df_ddi_val, df_ddi_test


df_ddi_train, df_ddi_val, df_ddi_test = load_data(raw_data_dir)

# calculate statistics
def calculate_stats(df_ddi, split):
    # implement this function to calculate the statistics
    # it is encouraged to print out the results

    # statistics for the dataset
    # (1) number of samples; (2) number of drugs [set for d1 and d2]; (3) number of drug interactions [type]
    num_samples = df_ddi.shape[0]
    num_drugs = len(set(df_ddi['d1'].values) | set(df_ddi['d2'].values))
    num_interactions = len(set(df_ddi['type'].values))
    
    print(f'Split: {split}')
    print(f'Number of samples: {num_samples}')
    print(f'Number of drugs: {num_drugs}')
    print(f'Number of interactions: {num_interactions}')
    print('*' * 50)


calculate_stats(df_ddi_train, 'train')
calculate_stats(df_ddi_val, 'val')
calculate_stats(df_ddi_test, 'test')



Split: train
Number of samples: 115185
Number of drugs: 1671
Number of interactions: 86
**************************************************
Split: val
Number of samples: 38348
Number of drugs: 1582
Number of interactions: 86
**************************************************
Split: test
Number of samples: 38337
Number of drugs: 1597
Number of interactions: 86
**************************************************


In [2]:
train_tup_ddi = [(h, t, r) for h, t, r in zip(df_ddi_train['d1'], df_ddi_train['d2'], df_ddi_train['type'])]
val_tup_ddi = [(h, t, r) for h, t, r in zip(df_ddi_val['d1'], df_ddi_val['d2'], df_ddi_val['type'])]
test_tup_ddi = [(h, t, r) for h, t, r in zip(df_ddi_test['d1'], df_ddi_test['d2'], df_ddi_test['type'])]

## Data 2: Protein-protein Interaction (PPI) Prediction

In [3]:
from torch_geometric.datasets import PPI
from torch_geometric.loader import DataLoader

def load_data(data_dir):
    path = data_dir
    train_dataset = PPI(path, split='train')
    val_dataset = PPI(path, split='val')
    test_dataset = PPI(path, split='test')

    return train_dataset, val_dataset, test_dataset

train_dataset_ppi, val_dataset_ppi, test_dataset_ppi = load_data('data/ppi')


# outut the statistics
def calculate_stats(dataset, split):
    # Initialize counters
    num_graphs = len(dataset)
    num_nodes = 0
    num_edges = 0

    # Loop through the dataset to accumulate statistics
    for data in dataset:
        num_nodes += data.num_nodes
        num_edges += data.num_edges
    
    # Since features and classes are consistent across the dataset, we can take them from the first graph
    num_features = dataset[0].num_features
    num_classes = dataset[0].y.size(1)  # Assuming the labels are one-hot encoded

    print(f'Split: {split}')
    print(f'Number of graphs: {num_graphs}')
    print(f'Number of nodes: {num_nodes}')
    print(f'Number of edges: {num_edges}')
    print(f'Number of features: {num_features}')
    print(f'Number of classes: {num_classes}')
    print('*' * 50)

calculate_stats(train_dataset_ppi, 'train')
calculate_stats(val_dataset_ppi, 'val')
calculate_stats(test_dataset_ppi, 'test')

Extracting data/ppi/ppi.zip
Processing...


Split: train
Number of graphs: 20
Number of nodes: 44906
Number of edges: 1226368
Number of features: 50
Number of classes: 121
**************************************************
Split: val
Number of graphs: 2
Number of nodes: 6514
Number of edges: 198920
Number of features: 50
Number of classes: 121
**************************************************
Split: test
Number of graphs: 2
Number of nodes: 5524
Number of edges: 161976
Number of features: 50
Number of classes: 121
**************************************************


Done!


## Data 3: PubMed Citation Network

In [4]:
from torch_geometric.datasets import CitationFull
import torch_geometric.transforms as T
import torch

device = 'cuda' if torch.cuda.is_available() else 'cpu'

transform = T.Compose([
    T.NormalizeFeatures(),
    T.ToDevice(device),
    T.RandomLinkSplit(num_val=0.05, num_test=0.1, is_undirected=True,
                      add_negative_train_samples=False),
])

def load_pubmed_data(data_dir):
    dataset = CitationFull(root=data_dir, name='PubMed', transform=transform)
    train_data, val_data, test_data = dataset[0]

    return train_data, val_data, test_data

train_data_pubmed, val_data_pubmed, test_data_pubmed = load_pubmed_data('data/pubmed')

def calculate_stats_citation(dataset, split):
    # For citation networks processed by RandomLinkSplit, we typically have one graph represented by edge splits.
    num_nodes = dataset.num_nodes  # Total number of nodes in the dataset
    num_edges = dataset['edge_index'].size(1) // 2  # Divided by 2 for undirected graphs
    num_features = dataset.num_features  # Number of features per node
    num_classes = dataset.y.unique().size(0)  # Assuming 'y' holds class labels for nodes
    # Print the statistics
    print(f'Split: {split}')
    print("Statistics for the citation network:")
    print(f"Number of graphs: 1")
    print(f"Number of nodes: {num_nodes}")
    print(f"Number of edges: {num_edges}")
    print(f"Number of features: {num_features}")
    print(f"Number of classes: {num_classes}")
    print('*' * 50)


calculate_stats_citation(train_data_pubmed, 'train')
calculate_stats_citation(val_data_pubmed, 'val')
calculate_stats_citation(test_data_pubmed, 'test')


Processing...
Done!


Split: train
Statistics for the citation network:
Number of graphs: 1
Number of nodes: 19717
Number of edges: 37676
Number of features: 500
Number of classes: 3
**************************************************
Split: val
Statistics for the citation network:
Number of graphs: 1
Number of nodes: 19717
Number of edges: 37676
Number of features: 500
Number of classes: 3
**************************************************
Split: test
Statistics for the citation network:
Number of graphs: 1
Number of nodes: 19717
Number of edges: 39892
Number of features: 500
Number of classes: 3
**************************************************


##   3.2 Model & Training & Evaluation
The model includes the model definitation which usually is a class, model training, and other necessary parts.
  * Model architecture: layer number/size/type, activation function, etc
  * Training objectives: loss function, optimizer, weight of each loss term, etc
  * Others: whether the model is pretrained, Monte Carlo simulation for uncertainty analysis, etc
  * The code of model should have classes of the model, functions of model training, model validation, etc.
  * If your model training is done outside of this notebook, please upload the trained model here and develop a function to load and test it.

1. Model architecture: \
(1) DDI: given a input triplet (drug A, drug B, relation), we use one GAT to encode both drugs and nn.Embeddings for relations. Then, we concat them and feed the concat tensor to a classifier. (2) PPI: Multi-layer GATs. (3) Link Prediction: Graph Autoencoders. More details can be found in the following code which is extremely clear.

2. Training objectives: \
(1) DDI: loss_func is -logsigmoid(pos_tri) - logsigmoid(-neg_tri). Adam optimizer. (2) PPI: BCE loss, Adam optimizer; (3) Link Prediciton: BCE loss, Adam optimizer;

3. Others: \
All the models in this project are trained from scratch.

4. Computational requirements: \
All the models can be run with a single 40G A6000.

5. Evaluation Metrics:
All the three tasks can be attributed to classification tasks. Thus, the metrics for evaluation are accuracy, precision, recall, F1 scores, and AUROC.

In [6]:
from tqdm.notebook import tqdm

def save_checkpoint(state, filename):
    if not os.path.exists('./ckpt'):
        os.makedirs('./ckpt')
    torch.save(state, f'./ckpt/{filename}.pth')

def load_checkpoint(model, optimizer, scheduler, filename):
    checkpoint = torch.load(f'./ckpt/{filename}.pth')
    model.load_state_dict(checkpoint['model_state_dict'])
    optimizer.load_state_dict(checkpoint['optimizer_state_dict'])
    if 'scheduler_state_dict' in checkpoint and scheduler:
        scheduler.load_state_dict(checkpoint['scheduler_state_dict'])
    epoch = checkpoint['epoch']
    return epoch

import re

def find_last_checkpoint_by_epoch(prefix, checkpoint_dir='ckpt'):
    # prefix = task name, ddi, ppi, PubMed-[model]
    pattern = re.compile(prefix + r"_e(\d{3}).pth")
    checkpoint_files = [f for f in os.listdir(checkpoint_dir) if os.path.isfile(os.path.join(checkpoint_dir, f)) and pattern.match(f)]
    epochs = [(int(pattern.search(f).group(1)), f) for f in checkpoint_files]
    if epochs:
        max_epoch_file = max(epochs, key=lambda x: x[0])[1]
        return os.path.join(checkpoint_dir, max_epoch_file)
    else:
        return None

## Task 1: Training GAT for Drug-drug Interaction
The input and output for DDI prediction is
* Input: Concat[Drug A, Drug B, relation]
* Output: Predicted Probability

In [26]:
from utils_ddi import DrugDataset, DrugDataLoader, TOTAL_ATOM_FEATS
import numpy as np
import torch
from sklearn import metrics
from datetime import datetime
import time
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F

# Hyperparameters
n_atom_feats = TOTAL_ATOM_FEATS
n_atom_hid = 64
rel_total = 86
lr = 1e-2
weight_decay = 5e-4
n_epochs = 300
neg_samples = 1
batch_size = 1024

train_data_ddi = DrugDataset(train_tup_ddi, ratio=1, neg_ent=1)
val_data_ddi = DrugDataset(val_tup_ddi, ratio=1, disjoint_split=False)
test_data_ddi = DrugDataset(test_tup_ddi, disjoint_split=False)

train_data_loader_ddi = DrugDataLoader(train_data_ddi, batch_size=batch_size, shuffle=True)
val_data_loader_ddi = DrugDataLoader(val_data_ddi, batch_size=batch_size *3)
test_data_loader_ddi = DrugDataLoader(test_data_ddi, batch_size=batch_size *3)

In [29]:
def do_compute(model, batch, device, training=True):
    '''
        *batch: (pos_tri, neg_tri)
        *pos/neg_tri: (batch_h, batch_t, batch_r)
    '''
    probas_pred, ground_truth = [], []
    pos_tri, neg_tri = batch
    
    pos_tri = [tensor.to(device=device) for tensor in pos_tri]
    p_score = model(pos_tri)
    probas_pred.append(torch.sigmoid(p_score.detach()).cpu())
    ground_truth.append(np.ones(len(p_score)))

    neg_tri = [tensor.to(device=device) for tensor in neg_tri]
    n_score = model(neg_tri)
    probas_pred.append(torch.sigmoid(n_score.detach()).cpu())
    ground_truth.append(np.zeros(len(n_score)))

    probas_pred = np.concatenate(probas_pred)
    ground_truth = np.concatenate(ground_truth)

    return p_score, n_score, probas_pred, ground_truth

def do_compute_metrics(probas_pred, target):

    pred = (probas_pred >= 0.5).astype(np.int32)

    acc = metrics.accuracy_score(target, pred)
    auc_roc = metrics.roc_auc_score(target, probas_pred)
    f1_score = metrics.f1_score(target, pred)

    p, r, t = metrics.precision_recall_curve(target, probas_pred)
    auc_prc = metrics.auc(r, p)

    return acc, auc_roc, auc_prc

def train(model, train_data_loader, val_data_loader, loss_fn,  optimizer, n_epochs, device, 
          scheduler=None, starting_epoch=1):
    print('Starting training at', datetime.today())
    pbar = tqdm(range(starting_epoch, n_epochs+1))
    for i in pbar:
        start = time.time()
        train_loss = 0
        train_loss_pos = 0
        train_loss_neg = 0
        val_loss = 0
        val_loss_pos = 0
        val_loss_neg = 0
        train_probas_pred = []
        train_ground_truth = []
        val_probas_pred = []
        val_ground_truth = []

        for batch in train_data_loader:
            model.train()
            p_score, n_score, probas_pred, ground_truth = do_compute(model, batch, device)
            train_probas_pred.append(probas_pred)
            train_ground_truth.append(ground_truth)
            loss, loss_p, loss_n = loss_fn(p_score, n_score)
            
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

            train_loss += loss.item() * len(p_score)
        train_loss /= len(train_data_ddi)

        with torch.no_grad():
            train_probas_pred = np.concatenate(train_probas_pred)
            train_ground_truth = np.concatenate(train_ground_truth)

            train_acc, train_auc_roc, train_auc_prc = do_compute_metrics(train_probas_pred, train_ground_truth)

            for batch in val_data_loader:
                model.eval()
                p_score, n_score, probas_pred, ground_truth = do_compute(model, batch, device)
                val_probas_pred.append(probas_pred)
                val_ground_truth.append(ground_truth)
                loss, loss_p, loss_n = loss_fn(p_score, n_score)
                val_loss += loss.item() * len(p_score)            

            val_loss /= len(val_data_ddi)
            val_probas_pred = np.concatenate(val_probas_pred)
            val_ground_truth = np.concatenate(val_ground_truth)
            val_acc, val_auc_roc, val_auc_prc = do_compute_metrics(val_probas_pred, val_ground_truth)
               
        if scheduler:
            # print('scheduling')
            scheduler.step()


        pbar.set_description(f'tr_loss: {train_loss:.4f}, val_loss: {val_loss:.4f},'
                   f' tr_acc: {train_acc:.4f}, val_acc: {val_acc:.4f},'
                   f' tr_roc: {train_auc_roc:.4f}, val_roc: {val_auc_roc:.4f},'
                   f' tr_aupr: {train_auc_prc:.4f}, val_aupr: {val_auc_prc:.4f}' )

        if i % 10 == 0 or i == n_epochs:  
            save_checkpoint({
                'epoch': i,
                'model_state_dict': model.state_dict(),
                'optimizer_state_dict': optimizer.state_dict(),
                'scheduler_state_dict': scheduler.state_dict() if scheduler else None, 
                'metrics': {'train_loss' : train_loss,
                           'val_loss' : val_loss,
                           'train_acc' : train_acc,
                           'val_acc' : val_acc,
                           'train_auc_roc' : train_auc_roc,
                           'val_auc_roc' : val_auc_roc,
                           'train_auprc' : train_auc_prc,
                           'val_auprc' : val_auc_prc},
            }, filename=f"ddi_e{str(i).zfill(3)}.pth")


class SigmoidLoss(nn.Module):
    def __init__(self, adv_temperature=None):
        super().__init__()
        self.adv_temperature = adv_temperature
    
    def forward(self, p_scores, n_scores):
        if self.adv_temperature:
            weights= F.softmax(self.adv_temperature * n_scores, dim=-1).detach()
            n_scores = weights * n_scores
        p_loss = - F.logsigmoid(p_scores).mean()
        n_loss = - F.logsigmoid(-n_scores).mean()
        
        return (p_loss + n_loss) / 2, p_loss, n_loss 

In [ ]:
from torch_geometric.nn import GATConv, global_mean_pool, LayerNorm
import torch.nn.functional as F


class DDIGAT(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.num_iterations = 86
        self.num_features = 55
        self.num_heads = 2
        self.initial_norm = LayerNorm(self.num_features)
        self.convA = GATConv(self.num_features, 256, heads=self.num_heads)
        self.convB = GATConv(self.num_features, 256, heads=self.num_heads)
        self.emb = torch.nn.Embedding(self.num_iterations, 256)
        self.lin = torch.nn.Linear((2 * self.num_heads + 1)* 256, 1)

    def forward(self, tri):
        # (pos_h_samples, pos_t_samples, pos_rels)
        head, tail, rel = tri
        head.x = self.initial_norm(head.x, head.batch)
        tail.x = self.initial_norm(tail.x, tail.batch)
        head = F.elu(self.convA(head.x, head.edge_index))
        tail = F.elu(self.convB(tail.x, tail.edge_index))
        head_emb = global_mean_pool(head, tri[0].batch)
        tail_emb = global_mean_pool(tail, tri[1].batch)
        rel = self.emb(rel)
        out = self.lin(torch.cat([head_emb, tail_emb, rel],dim=-1))
        return out

model = DDIGAT()
loss_func = SigmoidLoss()
optimizer = optim.Adam(model.parameters(), lr=lr, weight_decay=weight_decay)
scheduler = optim.lr_scheduler.LambdaLR(optimizer, lambda epoch: 0.96 ** (epoch))

device = 'cuda' if torch.cuda.is_available() else 'cpu'
model.to(device)

last_checkpoint = find_last_checkpoint_by_epoch('ckpt')

if last_checkpoint:
    last_epoch, last_loss = load_checkpoint(model, optimizer, scheduler, last_checkpoint)
    print(f"Resuming training from epoch {last_epoch + 1}")
    train(model, train_data_loader_ddi, val_data_loader_ddi, SigmoidLoss(), optimizer, n_epochs, device, scheduler, starting_epoch=last_epoch + 1)
else:
    print("No checkpoint found, starting training from scratch.")
    train(model, train_data_loader_ddi, val_data_loader_ddi, loss_func, optimizer, n_epochs, device, scheduler)

No checkpoint found, starting training from scratch.
Starting training at 2024-04-12 03:47:42.278611


  0%|          | 0/300 [00:00<?, ?it/s]

## Task 2: Training GAT for Protein-protein Interaction

In [7]:
from torch_geometric.nn import GATConv
import torch.nn.functional as F

train_loader_ppi = DataLoader(train_dataset_ppi, batch_size=1, shuffle=True)
val_loader_ppi = DataLoader(val_dataset_ppi, batch_size=2, shuffle=False)
test_loader_ppi = DataLoader(test_dataset_ppi, batch_size=2, shuffle=False)

class PPINet(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = GATConv(train_dataset_ppi.num_features, 256, heads=4)
        self.lin1 = torch.nn.Linear(train_dataset_ppi.num_features, 4 * 256)
        self.conv2 = GATConv(4 * 256, 256, heads=4)
        self.lin2 = torch.nn.Linear(4 * 256, 4 * 256)
        self.conv3 = GATConv(4 * 256, train_dataset_ppi.num_classes, heads=6,
                             concat=False)
        self.lin3 = torch.nn.Linear(4 * 256, train_dataset_ppi.num_classes)

    def forward(self, x, edge_index):
        x = F.elu(self.conv1(x, edge_index) + self.lin1(x))
        x = F.elu(self.conv2(x, edge_index) + self.lin2(x))
        x = self.conv3(x, edge_index) + self.lin3(x)
        return x

In [8]:
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model = PPINet().to(device)
loss_op = torch.nn.BCEWithLogitsLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.005)

from sklearn.metrics import f1_score


def train():
    model.train()

    total_loss = 0
    for data in train_loader_ppi:
        data = data.to(device)
        optimizer.zero_grad()
        loss = loss_op(model(data.x, data.edge_index), data.y)
        total_loss += loss.item() * data.num_graphs
        loss.backward()
        optimizer.step()

    return total_loss / len(train_loader_ppi.dataset)


@torch.no_grad()
def test(loader):
    model.eval()

    ys, preds = [], []
    for data in loader:
        ys.append(data.y)
        out = model(data.x.to(device), data.edge_index.to(device))
        preds.append((out > 0).float().cpu())

    y, pred = torch.cat(ys, dim=0).numpy(), torch.cat(preds, dim=0).numpy()
    return f1_score(y, pred, average='micro') if pred.sum() > 0 else 0



In [14]:
# import time

# times = []
# pbar = tqdm(range(1, 101))
# for epoch in pbar:
#     start = time.time()
#     loss = train()
#     val_f1 = test(val_loader_ppi)
#     test_f1 = test(test_loader_ppi)
#     if epoch % 10 == 0 or epoch == 100:  
#         save_checkpoint({
#             'epoch': epoch,
#             'model_state_dict': model.state_dict(),
#             'optimizer_state_dict': optimizer.state_dict(),
#             'metrics': {'train_loss' : loss,
#                        'val_f1' : val_f1,
#                        'test_f1' : test_f1},
#         }, filename=f"ppi_e{str(epoch).zfill(3)}.pth")

#     pbar.set_description(f'tr_loss: {loss:.4f}, val_f1: {val_f1:.4f}, test_f1: {test_f1:.4f}')
#     times.append(time.time() - start)
# print(f"Median time per epoch: {torch.tensor(times).median():.4f}s")

  0%|          | 0/100 [00:00<?, ?it/s]

Median time per epoch: 1.8779s


## Task 3: Training GAT for Link Prediction for PubMed Citation Network

In [19]:
from torch_geometric.utils import negative_sampling
from torch_geometric.nn import GATv2Conv, GCNConv

class GCNNet(torch.nn.Module):
    def __init__(self, in_channels, hidden_channels, out_channels):
        super().__init__()
        self.conv1 = GCNConv(in_channels, hidden_channels)
        self.conv2 = GCNConv(hidden_channels, out_channels)

    def encode(self, x, edge_index):
        x = self.conv1(x, edge_index).relu()
        return self.conv2(x, edge_index)

    def decode(self, z, edge_label_index):
        return (z[edge_label_index[0]] * z[edge_label_index[1]]).sum(dim=-1)

    def decode_all(self, z):
        prob_adj = z @ z.t()
        return (prob_adj > 0).nonzero(as_tuple=False).t()

class GATv2Net(torch.nn.Module):
    def __init__(self, in_channels, hidden_channels, out_channels, num_heads):
        super().__init__()
        self.conv1 = GATv2Conv(in_channels, hidden_channels, heads=num_heads, concat=False)
        self.conv2 = GATv2Conv(hidden_channels, out_channels, heads=num_heads, concat=False)
    
    def encode(self, x, edge_index):
        x = self.conv1(x, edge_index).relu()
        return self.conv2(x, edge_index)

    def decode(self, z, edge_label_index):
        return (z[edge_label_index[0]] * z[edge_label_index[1]]).sum(dim=-1)

    def decode_all(self, z):
        prob_adj = z @ z.t()
        return (prob_adj > 0).nonzero(as_tuple=False).t()
    

class GATv1Net(torch.nn.Module):
    def __init__(self, in_channels, hidden_channels, out_channels, num_heads):
        super().__init__()
        self.conv1 = GATConv(in_channels, hidden_channels, heads=num_heads, concat=False)
        self.conv2 = GATConv(hidden_channels, out_channels, heads=num_heads, concat=False)

    def encode(self, x, edge_index):
        x = self.conv1(x, edge_index).relu()
        return self.conv2(x, edge_index)

    def decode(self, z, edge_label_index):
        return (z[edge_label_index[0]] * z[edge_label_index[1]]).sum(dim=-1)

    def decode_all(self, z):
        prob_adj = z @ z.t()
        return (prob_adj > 0).nonzero(as_tuple=False).t()
   


In [20]:
from sklearn.metrics import roc_auc_score

optimizer = torch.optim.Adam(params=model.parameters(), lr=0.01)
criterion = torch.nn.BCEWithLogitsLoss()

def train():
    model.train()
    optimizer.zero_grad()
    z = model.encode(train_data_pubmed.x, train_data_pubmed.edge_index)

    # We perform a new round of negative sampling for every training epoch:
    neg_edge_index = negative_sampling(
        edge_index=train_data_pubmed.edge_index, num_nodes=train_data_pubmed.num_nodes,
        num_neg_samples=train_data_pubmed.edge_label_index.size(1), method='sparse')

    edge_label_index = torch.cat(
        [train_data_pubmed.edge_label_index, neg_edge_index],
        dim=-1,
    )
    edge_label = torch.cat([
        train_data_pubmed.edge_label,
        train_data_pubmed.edge_label.new_zeros(neg_edge_index.size(1))
    ], dim=0)

    out = model.decode(z, edge_label_index).view(-1)
    loss = criterion(out, edge_label)
    loss.backward()
    optimizer.step()
    return loss


@torch.no_grad()
def test(data):
    model.eval()
    z = model.encode(data.x, data.edge_index)
    out = model.decode(z, data.edge_label_index).view(-1).sigmoid()
    return roc_auc_score(data.edge_label.cpu().numpy(), out.cpu().numpy())


In [22]:
for model_name in ['GCN','GAT','GAT-multi','GATv2','GATv2-multi']:
    print(model_name)
    
    if model_name == 'GCN':
        model = GCNNet(train_data_pubmed.num_features, 128, 64).to(device)
    elif model_name == 'GAT':
        model = GATv1Net(train_data_pubmed.num_features, 128, 64, 1).to(device)
    elif model_name == 'GAT-multi':
        model = GATv1Net(train_data_pubmed.num_features, 128, 64, 4).to(device)
    elif model_name == 'GATv2':
        model = GATv2Net(train_data_pubmed.num_features, 128, 64, 1).to(device)
    elif model_name == 'GATv2-multi':
        model = GATv2Net(train_data_pubmed.num_features, 128, 64, 4).to(device)

    
    best_val_auc = final_test_auc = 0
    pbar = tqdm(range(1, 101))
    for epoch in pbar:
        loss = train()
        val_auc = test(val_data_pubmed)
        test_auc = test(test_data_pubmed)
        if val_auc > best_val_auc:
            best_val_auc = val_auc
            final_test_auc = test_auc
            save_checkpoint({
                'epoch': epoch,
                'model_state_dict': model.state_dict(),
                'optimizer_state_dict': optimizer.state_dict(),
                'metrics': {'train_loss' : loss,
                           'val_auc' : val_auc,
                           'test_auc' : test_auc},
            }, filename=f"PubMed_{model_name}_best.pth")
        if epoch % 10 == 0 or epoch == 100:  
            save_checkpoint({
                'epoch': epoch,
                'model_state_dict': model.state_dict(),
                'optimizer_state_dict': optimizer.state_dict(),
                'metrics': {'train_loss' : loss,
                           'val_auc' : val_auc,
                           'test_auc' : test_auc},
            }, filename=f"PubMed_{model_name}_e{str(epoch).zfill(3)}.pth")
        pbar.set_description(f'tr_loss: {loss:.4f}, val_auc: {val_auc:.4f}, test_auc: {test_auc:.4f}')
    
    print(f'{model_name} Final Test: {final_test_auc:.4f}')

GCN


  0%|          | 0/100 [00:00<?, ?it/s]

GCN Final Test: 0.9492
GAT


  0%|          | 0/100 [00:00<?, ?it/s]

GAT Final Test: 0.8959
GAT-multi


  0%|          | 0/100 [00:00<?, ?it/s]

GAT-multi Final Test: 0.7966
GATv2


  0%|          | 0/100 [00:00<?, ?it/s]

GATv2 Final Test: 0.8736
GATv2-multi


  0%|          | 0/100 [00:00<?, ?it/s]

GATv2-multi Final Test: 0.7941


# 3.3 Results
In this section, you should finish training your model training or loading your trained model. That is a great experiment! You should share the results with others with necessary metrics and figures.

Please test and report results for all experiments that you run with:

*   specific numbers (accuracy, AUC, RMSE, etc)
*   figures (loss shrinkage, outputs from GAN, annotation or label of sample pictures, etc)


### Task 1: Drug-drug Interaction
|Model|Accuracy|AUROC|F1|Precision|Recall|
|--|--|--|--|--|--|
|GAT| | | | | |


### Task 2: Protein-protein Interaction
|Model|Accuracy|AUROC|F1|Precision|Recall|
|--|--|--|--|--|--|
|GAT| | | | | |

### Task 3: Link Prediciton
|Model|Accuracy|AUROC|F1|Precision|Recall|
|--|--|--|--|--|--|
|GAT (single-head)| | 0.8959| | | |
|GAT (multi-head)| | 0.7966| | | |
|GATv2 (single-head)| |0.7941 | | | |
|GATv2 (multi-head)| | 0.8736| | | |
|GCN| | 0.9492| | | |


#### 3.3.1 Analysis
From the results for link prediction, GCN shows the best performance. GAT (multi-head) is better than GAT (single head). GATv2 is worse than GAT. Although all of the above results are beyond our expectation, they does obey the "No free lunch" rule. 

In [ ]:
# metrics to evaluate my model

# plot figures to better show the results

# it is better to save the numbers and figures for your presentation.

## 3.4 Model comparison

In [ ]:
# compare you model with others
# you don't need to re-run all other experiments, instead, you can directly refer the metrics/numbers in the paper

# 4 Plan
* Before April 25th, finish all the experiments and figures, as well as the analysis.

# 4 Discussion

In this section,you should discuss your work and make future plan. The discussion should address the following questions:
  * Make assessment that the paper is reproducible or not.
  * Explain why it is not reproducible if your results are kind negative.
  * Describe “What was easy” and “What was difficult” during the reproduction.
  * Make suggestions to the author or other reproducers on how to improve the reproducibility.
  * What will you do in next phase.



In [ ]:
# no code is required for this section
'''
if you want to use an image outside this notebook for explanaition,
you can read and plot it here like the Scope of Reproducibility
'''

# References

Shaked Brody, Uri Alon, and Eran Yahav. How attentive are graph attention networks? In The Tenth
International Conference on Learning Representations, ICLR 2022, Virtual Event, April 25-29,
2022. OpenReview.net, 2022. URL https://openreview.net/forum?id=F72ximsx7C1

Lin, J., Wu, L., Zhu, J., Liang, X., Xia, Y., Xie, S., ... & Liu, T. Y. (2023). R2-DDI: relation-aware feature refinement for drug–drug interaction prediction. Briefings in Bioinformatics, 24(1), bbac576.

Petar Velickovic, Guillem Cucurull, Arantxa Casanova, Adriana Romero, Pietro Liò, and Yoshua
Bengio. Graph attention networks. In 6th International Conference on Learning Representations,
ICLR 2018, Vancouver, BC, Canada, April 30 - May 3, 2018, Conference Track Proceedings.
OpenReview.net, 2018. URL https://openreview.net/forum?id=rJXMpikCZ.

